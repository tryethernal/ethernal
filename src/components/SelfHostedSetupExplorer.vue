<template>
  <v-alert v-if="error" type="error" class="mb-4">{{ error }}</v-alert>
  <v-form @submit.prevent="submit" v-model="valid" ref="form" class="mt-2">
    <v-text-field
      v-model="name"
      label="Explorer Name"
      :rules="nameRules"
      required
      class="mb-4"
    />
    <v-text-field
      v-model="rpcUrl"
      label="RPC URL"
      :rules="rpcRules"
      required
      class="mb-4"
    />
    <v-btn :loading="props.loading || submitting" color="primary" type="submit" :disabled="!valid || props.loading || submitting" block>
      Create Explorer
    </v-btn>
  </v-form>
</template>

<script setup>
import { ref, inject } from 'vue';

const props = defineProps({
  loading: {
    type: Boolean,
    default: false
  }
});
const emit = defineEmits(['explorer-created']);

const $server = inject('$server');

const name = ref('');
const rpcUrl = ref('');
const error = ref('');
const valid = ref(false);
const form = ref(null);
const submitting = ref(false);
const nameRules = [
  v => !!v || 'Name is required',
];
const rpcRules = [
  v => !!v || 'RPC URL is required',
  v => /^https?:\/\//.test(v) || 'Must be a valid URL',
];

async function submit() {
  error.value = '';
  if (!valid.value) return;
  submitting.value = true;
  try {
    const { data: explorer } = await $server.createExplorerFromOptions(name.value, rpcUrl.value);
    emit('explorer-created', explorer);
  } catch (e) {
    error.value = e?.response?.data || e?.message || 'Failed to create explorer.';
  } finally {
    submitting.value = false;
  }
}
</script>
