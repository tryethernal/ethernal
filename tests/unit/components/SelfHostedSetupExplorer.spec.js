import SelfHostedSetupExplorer from '@/components/SelfHostedSetupExplorer.vue';

const stubs = [];

describe('SelfHostedSetupExplorer.vue', () => {
    it('validates required fields', async () => {
        const wrapper = mount(SelfHostedSetupExplorer, {
            global: {
                stubs,
                provide: {
                    $server: {
                        getCurrentUser: vi.fn().mockResolvedValue({ data: { id: 1 } }),
                        createExplorerFromOptions: vi.fn().mockResolvedValue({ data: { id: 42, slug: 'slug' } })
                    }
                }
            }
        });
        await wrapper.find('form').trigger('submit.prevent');
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('shows error on failed explorer creation', async () => {
        const createExplorerFromOptions = vi.fn().mockRejectedValueOnce({ message: 'fail' });
        const wrapper = mount(SelfHostedSetupExplorer, {
            global: {
                stubs,
                provide: {
                    $server: {
                        getCurrentUser: vi.fn().mockResolvedValue({ data: { id: 1 } }),
                        createExplorerFromOptions
                    }
                }
            }
        });
        await wrapper.find('input[type=text]').setValue('My Explorer');
        await wrapper.findAll('input[type=text]')[1].setValue('http://localhost:8545');
        await wrapper.find('form').trigger('submit.prevent');
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('emits explorer-created on success', async () => {
        const getCurrentUser = vi.fn().mockResolvedValue({ data: { id: 1 } });
        const createExplorerFromOptions = vi.fn().mockResolvedValue({ data: { id: 42, slug: 'slug' } });
        const wrapper = mount(SelfHostedSetupExplorer, {
            global: {
                stubs,
                provide: {
                    $server: {
                        getCurrentUser,
                        createExplorerFromOptions
                    }
                }
            }
        });
        await wrapper.find('input[type=text]').setValue('My Explorer');
        await wrapper.findAll('input[type=text]')[1].setValue('http://localhost:8545');
        await wrapper.find('form').trigger('submit.prevent');
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });
});
