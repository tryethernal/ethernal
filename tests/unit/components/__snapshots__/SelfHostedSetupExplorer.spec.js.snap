// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SelfHostedSetupExplorer.vue > emits explorer-created on success 1`] = `
"<!--v-if-->
<form class="v-form mt-2" novalidate="">
  <div class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-input--dirty v-text-field mb-4">
    <!---->
    <div class="v-input__control">
      <div class="v-field v-field--active v-field--center-affix v-field--dirty v-field--variant-filled v-theme--light v-locale--is-ltr">
        <div class="v-field__overlay"></div>
        <div class="v-field__loader">
          <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
            <!---->
            <div class="v-progress-linear__background"></div>
            <div class="v-progress-linear__buffer" style="width: 0%;"></div>
            <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
              <div class="v-progress-linear__indeterminate">
                <div class="v-progress-linear__indeterminate long"></div>
                <div class="v-progress-linear__indeterminate short"></div>
              </div>
            </transition-stub>
            <!---->
          </div>
        </div>
        <!---->
        <div class="v-field__field" data-no-activator=""><label class="v-label v-field-label v-field-label--floating" aria-hidden="true" for="input-0">
            <!---->Explorer Name
          </label><label class="v-label v-field-label" for="input-0">
            <!---->Explorer Name
          </label>
          <!----><input size="1" type="text" id="input-0" aria-describedby="input-0-messages" required="" class="v-field__input" value="My Explorer">
          <!---->
        </div>
        <!---->
        <!---->
        <div class="v-field__outline">
          <!---->
          <!---->
        </div>
      </div>
    </div>
    <!---->
    <div id="input-0-messages" class="v-input__details" role="alert" aria-live="polite">
      <transition-group-stub name="slide-y-transition" tag="div" appear="false" persisted="false" css="true" class="v-messages">
        <!---->
      </transition-group-stub>
      <!---->
    </div>
  </div>
  <div class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-input--dirty v-text-field mb-4">
    <!---->
    <div class="v-input__control">
      <div class="v-field v-field--active v-field--center-affix v-field--dirty v-field--variant-filled v-theme--light v-locale--is-ltr">
        <div class="v-field__overlay"></div>
        <div class="v-field__loader">
          <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
            <!---->
            <div class="v-progress-linear__background"></div>
            <div class="v-progress-linear__buffer" style="width: 0%;"></div>
            <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
              <div class="v-progress-linear__indeterminate">
                <div class="v-progress-linear__indeterminate long"></div>
                <div class="v-progress-linear__indeterminate short"></div>
              </div>
            </transition-stub>
            <!---->
          </div>
        </div>
        <!---->
        <div class="v-field__field" data-no-activator=""><label class="v-label v-field-label v-field-label--floating" aria-hidden="true" for="input-2">
            <!---->RPC URL
          </label><label class="v-label v-field-label" for="input-2">
            <!---->RPC URL
          </label>
          <!----><input size="1" type="text" id="input-2" aria-describedby="input-2-messages" required="" class="v-field__input" value="http://localhost:8545">
          <!---->
        </div>
        <!---->
        <!---->
        <div class="v-field__outline">
          <!---->
          <!---->
        </div>
      </div>
    </div>
    <!---->
    <div id="input-2-messages" class="v-input__details" role="alert" aria-live="polite">
      <transition-group-stub name="slide-y-transition" tag="div" appear="false" persisted="false" css="true" class="v-messages">
        <!---->
      </transition-group-stub>
      <!---->
    </div>
  </div><button type="submit" class="v-btn v-btn--block v-btn--elevated v-theme--light bg-primary v-btn--density-default v-btn--size-default v-btn--variant-elevated"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
    <!----><span class="v-btn__content" data-no-activator=""> Create Explorer </span>
    <!---->
    <!---->
  </button>
</form>"
`;

exports[`SelfHostedSetupExplorer.vue > shows error on failed explorer creation 1`] = `
"<!--v-if-->
<form class="v-form mt-2" novalidate="">
  <div class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-input--dirty v-text-field mb-4">
    <!---->
    <div class="v-input__control">
      <div class="v-field v-field--active v-field--center-affix v-field--dirty v-field--variant-filled v-theme--light v-locale--is-ltr">
        <div class="v-field__overlay"></div>
        <div class="v-field__loader">
          <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
            <!---->
            <div class="v-progress-linear__background"></div>
            <div class="v-progress-linear__buffer" style="width: 0%;"></div>
            <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
              <div class="v-progress-linear__indeterminate">
                <div class="v-progress-linear__indeterminate long"></div>
                <div class="v-progress-linear__indeterminate short"></div>
              </div>
            </transition-stub>
            <!---->
          </div>
        </div>
        <!---->
        <div class="v-field__field" data-no-activator=""><label class="v-label v-field-label v-field-label--floating" aria-hidden="true" for="input-0">
            <!---->Explorer Name
          </label><label class="v-label v-field-label" for="input-0">
            <!---->Explorer Name
          </label>
          <!----><input size="1" type="text" id="input-0" aria-describedby="input-0-messages" required="" class="v-field__input" value="My Explorer">
          <!---->
        </div>
        <!---->
        <!---->
        <div class="v-field__outline">
          <!---->
          <!---->
        </div>
      </div>
    </div>
    <!---->
    <div id="input-0-messages" class="v-input__details" role="alert" aria-live="polite">
      <transition-group-stub name="slide-y-transition" tag="div" appear="false" persisted="false" css="true" class="v-messages">
        <!---->
      </transition-group-stub>
      <!---->
    </div>
  </div>
  <div class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-input--dirty v-text-field mb-4">
    <!---->
    <div class="v-input__control">
      <div class="v-field v-field--active v-field--center-affix v-field--dirty v-field--variant-filled v-theme--light v-locale--is-ltr">
        <div class="v-field__overlay"></div>
        <div class="v-field__loader">
          <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
            <!---->
            <div class="v-progress-linear__background"></div>
            <div class="v-progress-linear__buffer" style="width: 0%;"></div>
            <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
              <div class="v-progress-linear__indeterminate">
                <div class="v-progress-linear__indeterminate long"></div>
                <div class="v-progress-linear__indeterminate short"></div>
              </div>
            </transition-stub>
            <!---->
          </div>
        </div>
        <!---->
        <div class="v-field__field" data-no-activator=""><label class="v-label v-field-label v-field-label--floating" aria-hidden="true" for="input-2">
            <!---->RPC URL
          </label><label class="v-label v-field-label" for="input-2">
            <!---->RPC URL
          </label>
          <!----><input size="1" type="text" id="input-2" aria-describedby="input-2-messages" required="" class="v-field__input" value="http://localhost:8545">
          <!---->
        </div>
        <!---->
        <!---->
        <div class="v-field__outline">
          <!---->
          <!---->
        </div>
      </div>
    </div>
    <!---->
    <div id="input-2-messages" class="v-input__details" role="alert" aria-live="polite">
      <transition-group-stub name="slide-y-transition" tag="div" appear="false" persisted="false" css="true" class="v-messages">
        <!---->
      </transition-group-stub>
      <!---->
    </div>
  </div><button type="submit" class="v-btn v-btn--block v-btn--elevated v-theme--light bg-primary v-btn--density-default v-btn--size-default v-btn--variant-elevated"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
    <!----><span class="v-btn__content" data-no-activator=""> Create Explorer </span>
    <!---->
    <!---->
  </button>
</form>"
`;

exports[`SelfHostedSetupExplorer.vue > validates required fields 1`] = `
"<!--v-if-->
<form class="v-form mt-2" novalidate="">
  <div class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-input--error v-text-field mb-4">
    <!---->
    <div class="v-input__control">
      <div class="v-field v-field--center-affix v-field--error v-field--variant-filled v-theme--light v-locale--is-ltr">
        <div class="v-field__overlay"></div>
        <div class="v-field__loader">
          <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
            <!---->
            <div class="v-progress-linear__background bg-error"></div>
            <div class="v-progress-linear__buffer bg-error" style="width: 0%;"></div>
            <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
              <div class="v-progress-linear__indeterminate">
                <div class="v-progress-linear__indeterminate long bg-error"></div>
                <div class="v-progress-linear__indeterminate short bg-error"></div>
              </div>
            </transition-stub>
            <!---->
          </div>
        </div>
        <!---->
        <div class="v-field__field" data-no-activator=""><label class="v-label v-field-label v-field-label--floating" aria-hidden="true" for="input-0">
            <!---->Explorer Name
          </label><label class="v-label v-field-label" for="input-0">
            <!---->Explorer Name
          </label>
          <!----><input size="1" type="text" id="input-0" aria-describedby="input-0-messages" required="" class="v-field__input" value="">
          <!---->
        </div>
        <!---->
        <!---->
        <div class="v-field__outline">
          <!---->
          <!---->
        </div>
      </div>
    </div>
    <!---->
    <div id="input-0-messages" class="v-input__details" role="alert" aria-live="polite">
      <transition-group-stub name="slide-y-transition" tag="div" appear="false" persisted="false" css="true" class="v-messages">
        <div class="v-messages__message">Name is required</div>
      </transition-group-stub>
      <!---->
    </div>
  </div>
  <div class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-text-field mb-4">
    <!---->
    <div class="v-input__control">
      <div class="v-field v-field--center-affix v-field--variant-filled v-theme--light v-locale--is-ltr">
        <div class="v-field__overlay"></div>
        <div class="v-field__loader">
          <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
            <!---->
            <div class="v-progress-linear__background"></div>
            <div class="v-progress-linear__buffer" style="width: 0%;"></div>
            <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
              <div class="v-progress-linear__indeterminate">
                <div class="v-progress-linear__indeterminate long"></div>
                <div class="v-progress-linear__indeterminate short"></div>
              </div>
            </transition-stub>
            <!---->
          </div>
        </div>
        <!---->
        <div class="v-field__field" data-no-activator=""><label class="v-label v-field-label v-field-label--floating" aria-hidden="true" for="input-2">
            <!---->RPC URL
          </label><label class="v-label v-field-label" for="input-2">
            <!---->RPC URL
          </label>
          <!----><input size="1" type="text" id="input-2" aria-describedby="input-2-messages" required="" class="v-field__input" value="">
          <!---->
        </div>
        <!---->
        <!---->
        <div class="v-field__outline">
          <!---->
          <!---->
        </div>
      </div>
    </div>
    <!---->
    <div id="input-2-messages" class="v-input__details" role="alert" aria-live="polite">
      <transition-group-stub name="slide-y-transition" tag="div" appear="false" persisted="false" css="true" class="v-messages">
        <!---->
      </transition-group-stub>
      <!---->
    </div>
  </div><button type="submit" class="v-btn v-btn--block v-btn--disabled v-theme--light bg-primary v-btn--density-default v-btn--size-default v-btn--variant-elevated" disabled=""><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
    <!----><span class="v-btn__content" data-no-activator=""> Create Explorer </span>
    <!---->
    <!---->
  </button>
</form>"
`;
